File: CSharp/ImageProcessing/ImageProcessing/ColorReversal.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;

/// <summary>
/// 色反転のロジック
/// </summary>
class ColorReversal
{
    /// <summary>
    /// コンストラクタ
    /// </summary>
    public ColorReversal()
    {
    }

    /// <summary>
    /// デスクトラクタ
    /// </summary>
    ~ColorReversal()
    {
    }

    /// <summary>
    /// 色反転の実行
    /// </summary>
    /// <param name="bitmap">ビットマップ</param>
    /// <param name="token">キャンセルトークン</param>
    /// <returns>実行結果 成功/失敗</returns>
    public bool ImageProcessing(ref Bitmap bitmap, CancellationToken token)
    {
        bool result = true;

        int widthSize = bitmap.Width;
        int heightSize = bitmap.Height;

        var bitmapData = bitmap.LockBits(new Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);

        int indexWidth;
        int indexHeight;

        unsafe
        {
            for (indexHeight = 0; indexHeight < heightSize; indexHeight++)
            {
                if (token.IsCancellationRequested)
                {
                    result = false;
                    break;
                }

                for (indexWidth = 0; indexWidth < widthSize; indexWidth++)
                {
                    if (token.IsCancellationRequested)
                    {
                        result = false;
                        break;
                    }

                    byte* pixel = (byte*)bitmapData.Scan0 + indexHeight * bitmapData.Stride + indexWidth * 4;

                    pixel[(int)ComInfo.Pixel.B] = (byte)(255 - pixel[(int)ComInfo.Pixel.B]);
                    pixel[(int)ComInfo.Pixel.G] = (byte)(255 - pixel[(int)ComInfo.Pixel.G]);
                    pixel[(int)ComInfo.Pixel.R] = (byte)(255 - pixel[(int)ComInfo.Pixel.R]);
                }
            }
            bitmap.UnlockBits(bitmapData);
        }

        return result;
    }
}

File: Vb/ImageProcessing/ImageProcessing/ColorReversal.vb
Imports System.Threading
Imports System.Runtime.InteropServices.Marshal
Imports System.Drawing.Imaging

''' <summary>
''' 色反転のロジック
''' </summary>
Public Class ColorReversal
    ''' <summary>
    ''' コンストラクタ
    ''' </summary>
    Public Sub New()
    End Sub

    ''' <summary>
    ''' デスクトラクタ
    ''' </summary>
    Protected Overrides Sub Finalize()
        MyBase.Finalize()
    End Sub

    ''' <summary>
    ''' 色反転の実行
    ''' </summary>
    ''' <param name="bitmap">ビットマップ</param>
    ''' <param name="token">キャンセルトークン</param>
    ''' <returns>実行結果 成功/失敗</returns>
    Public Function ImageProcessing(ByRef bitmap As Bitmap, token As CancellationToken) As Boolean
        Dim result = True

        Dim widthSize = bitmap.Width
        Dim heightSize = bitmap.Height

        Dim bitmapData = bitmap.LockBits(New Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

        Dim indexWidth As Integer
        Dim indexHeight As Integer

        For indexHeight = 0 To heightSize - 1 Step 1
            If (token.IsCancellationRequested = True) Then
                result = False
                Exit For
            End If

            For indexWidth = 0 To widthSize - 1 Step 1
                If (token.IsCancellationRequested = True) Then
                    result = False
                    Exit For
                End If

                Dim position = indexHeight * bitmapData.Stride + indexWidth * 4
                Dim pixelB = ReadByte(bitmapData.Scan0, position + ComInfo.Pixel.B)
                Dim pixelG = ReadByte(bitmapData.Scan0, position + ComInfo.Pixel.G)
                Dim pixelR = ReadByte(bitmapData.Scan0, position + ComInfo.Pixel.R)

                WriteByte(bitmapData.Scan0, position + ComInfo.Pixel.B, CByte(255 - pixelB))
                WriteByte(bitmapData.Scan0, position + ComInfo.Pixel.G, CByte(255 - pixelG))
                WriteByte(bitmapData.Scan0, position + ComInfo.Pixel.R, CByte(255 - pixelR))
            Next
        Next
        bitmap.UnlockBits(bitmapData)

        Return result
    End Function
End Class